tripy/tripy/frontend/module/module.py

```diff
@@ -17,21 +17,47 @@
 
 import copy
 import operator
-from typing import Any, Dict, Iterator, List, Set, Tuple, TypeVar, Union
+from typing import Any, Dict, Iterator, List, Set, Tuple, Union
 
-from tripy import export
+from tripy import export, utils
 from tripy.common.exception import raise_error
-from tripy.frontend.module.parameter import Parameter
+from tripy.frontend.module.parameter import DefaultParameter
+from tripy.frontend.tensor import Tensor
+from tripy.function_registry import type_str_from_arg
 from tripy.logging import logger
 
-T = TypeVar("T")
-
 
 def _check_param_compatible(original_param, new_param, param_name):
-    if not isinstance(original_param, Parameter):
+    # We want to check the incoming parameter type even when the original parameter is not a tensor.
+    if not isinstance(new_param, Tensor):
+        raise_error(
+            "Unrecognized type for module parameter.",
+            f"Expected a tensor for parameter: '{param_name}', but got: {type_str_from_arg(new_param)}",
+        )
+
+    if not isinstance(original_param, Tensor):
+        # Allow values to be initialized to non-tensor types and changed later.
```

Comment:
When do we set it to a non-tensor type? Only for the initial value? (Is it `None`?) Seems like odd to have an exception to the invariant.

Reply:
In Tripy, only when it's `None` (before the attribute has been set for the first time). But we don't want to be unnecessarily restrictive in user code. e.g. there's no reason to disallow:
```py
module.x = "example"
module.x = tp.Tensor(...)
```

Reply:
That's true, perhaps we might find uses for non-tensor-valued properties.